java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageHolder.java

```diff
@@ -0,0 +1,30 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.ipc.message;
+
+import io.netty.buffer.ArrowBuf;
+import org.apache.arrow.flatbuf.Message;
+
+/**
+ * Class to hold a Message and body when reading messages through a MessageChannelReader.
+ */
+public class MessageHolder {
```

Comment:
I am a bit confused about `MessageHolder` and `MessageChannelResult`, can you explain a bit what both are for and why do we need both?

Reply:
If we have a `MessageChannelReader`, I think that there should a single public method that will read a message + body. `MessageHolder` contains the message and body, whether this is passed by reference or returned by `readNext` doesn't matter to me.

`MessageChannelResult` is the returned object for the low-level function `readMessage` that just reads a flatbuffer message. This class has the message plus the raw data buffer used during reading. The raw data is useful for my case, but probably not most cases, which is why I kept it only in this low-level function.

Reply:
I see. `MessageHolder` seems fine to me because it's just contains two parts (header + body) for a message. `MessageChannelResult` does feel a bit arbitrary to me. Does `Message.getByteBuffer()` doesn't do what you want?

Reply:
`Message.getByteBuffer()` returns the same buffer right now because of the way it's created, but then I would be reliant on that assumption. If something changed internally in this function, then this might not be true. So by returning the raw byte buffer of the message makes a guarantee of the contents.

Reply:
Sorry for the blocking the PR at such time...But want to make sure we don't introduce something unnecessary here...

Reply:
> Do you mean that it's possible that Message.getByteBuffer() has different content than in MessageChannelResult.messageBuffer

I'm saying from what I can tell, there is no guarantee that `Message.getByteBuffer()` only contains the message bytes, and it's possible there could be additional bytes before or after. When the `Message` is created, they do some accounting with the buffer position and that is not public.

Reply:
Fair enough. I did some googling but didn't find a good answer either.

Reply:
@BryanCutler @icexelloss see my comments above; the lack of clarity in the comments to help distinguish the serialized metadata from the message body probably fueled this long comment thread